### ISSUE TITLE
Color artifacts using wide-gamut profiles

### ISSUE THREAD

**Issue by pi99y** [Beep6581/RawTherapee#4310]
Version: rawtherapee-unstable/xenial,now 1:5.3-git20180115-732~ubuntu16.04.1 amd64

When converting TIFF image to JPG the (blown) highlights get colored in an unpleasant way.
Highlights look normal while editing in RT. 
(see screenshot below)

I just noticed this with the latest update and not before (but I might have missed it)...

![highlights](https://user-images.githubusercontent.com/10639257/35009846-d6c30588-faf8-11e7-8572-df2ff854f789.png)

TIFF + pp3: https://filebin.net/y07bffs8hnhjzdp2/tiff_highlights.zip (11MB)

Thanks!

**Comment by agriggio** [comment]
hi @pi99y,
I cannot reproduce here. Can you post the details of your RT version (in preferences -> about -> version) please?

**Comment by Beep6581** [comment]
Cannot reproduce.
Have you made sure your color management settings are identical in both programs, including monitor profile and rendering intent? Any reason in particular you're using the AdobeRGB colorspace on an 8-bit image?

**Comment by pi99y** [comment]
Hi, thanks for the response guys...

@agriggio I'd love to report more details about the version but as reported in issue #3513 I (still) cannot. (Info not present)
Since I've updated to "rawtherapee-unstable/xenial,now 1:5.3-git20180117-737~ubuntu16.04.1 amd64 [installed]" (info from apt) and the problem persists.

@Beep6581
Both programs are set to use system's monitor profile and rendering is set to relative colorimetric. Also I see the same shift if I open the output JPG on my Android phone or Windows.

Link to output JPG as produced by my machine: https://filebin.net/zoq48deni35q8z9l/201712XX-201801XX_agfa-vista400_20_cleaned_rt.jpg (You see the shift?)

AdobeRGB is what scanner produces and the TIFFs are usually 16bit, looks like I saved to 8bit by mistake. Later I send these to print shop so I just keep 'em AdobeRGB. Any recommendations here?

**Comment by pi99y** [comment]
To confirm nothing is wrong with my "setup" I processed the TIFF/pp3 combo in an old version I compiled myself awhile ago, and there is no problem there (cant reproduce).
.
The old, working version is (just in case):
```
Branch: master
Version: 4.2.1208
Changeset: 3cc8438cf90a3001c2f01d44e2962c0c2c8dc262
Compiler: cc 5.4.0
Processor: x86_64
System: Linux
Bit depth: 64 bits
Gtkmm: V2.24.4
Build type: release
Build flags: -std=c++11 -Wno-deprecated-declarations -Wno-unused-result -std=gnu++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -O3 -DNDEBUG
Link flags:  -march=native
OpenMP support: ON
MMAP support: ON
```
So I think it's really the new version's problem.
As noted I get new versions from "ppa:dhor/myway".

What else can I do to make finding the bug easier?
Thanks!

**Comment by Benitoite** [comment]
Do you have this file?
/usr/share/doc/rawtherapee/AboutThisBuild.txt

fyi @salvadhor I'm looking at the [buildlog](https://launchpadlibrarian.net/353844914/buildlog_ubuntu-artful-amd64.rawtherapee-unstable_1%3A5.3-git20180117-737~ubuntu17.10.1_BUILDING.txt.gz), notice the useful info is stripped by the recipe. 
```
[  0%] Creating AboutThisBuild.txt and other version-dependent files
/usr/bin/cmake -DPROJECT_SOURCE_DIR:STRING=/<<PKGBUILDDIR>> -DCACHE_NAME_SUFFIX:STRING=5-dev -DPROC_LABEL:STRING="x86_64" -DPROC_BIT_DEPTH:STRING="64 bits" -DBUILD_TYPE:STRING=Release -DGTKMM_VERSION:STRING=3.22.2 -DOPTION_OMP:STRING=ON -DWITH_MYFILE_MMAP:STRING=ON -DLENSFUN_VERSION:STRING=0.3.2.0 -DSYSTEM:STRING=Linux -DCXX_FLAGS:STRING=-std=c++11\ -O3\ -std=c++11\ \ -Werror=unused-label\ -fopenmp\ -Werror=unknown-pragmas\ -Wall\ -Wno-unused-result\ -Wno-deprecated-declarations\ \ -ftree-vectorize -DLFLAGS:STRING=\ \  -DCOMPILER_INFO:STRING=cc\ 7.2.0 -P /<<PKGBUILDDIR>>/UpdateInfo.cmake
-- git command found: /usr/bin/git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
-- Git checkout information:
-- 	Commit description:	
-- 	Branch:			
-- 	Commit:			
-- 	Commit date:		
-- 	Commits since tag:	
-- 	Commits since branch:	
-- 	Version (unreliable):	0.0.0
-- CACHE_NAME_SUFFIX is "5-dev"
```

With the info it might be possible to notice a bug by rolling back to a previous build.

**Comment by Beep6581** [comment]
@pi99y can you reproduce this on any other images?
Please upload your `options` file.

**Comment by pi99y** [comment]
@Benitoite Sure, but like I said, the version and branch is absent...

AboutThisBuild.txt (from "/usr/share/rawtherapee/", another dir exists "/usr/share/doc/rawtherapee-unstable/" but does not include this file)

```
Version: 
Branch: 
Commit: 
Commit date: 
Compiler: x86_64-linux-gnu-gcc 5.4.0
Processor: x86_64
System: Linux
Bit depth: 64 bits
Gtkmm: V3.18.0
Lensfun: V0.2.8.0
Build type: Release
Build flags: -std=c++11 -O3 -std=c++11  -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations  -ftree-vectorize
Link flags:   
OpenMP support: ON
MMAP support: ON
```

@Beep6581 Yes, I can reproduce other images with this. Will upload shortly.
Which is the "options" file?

**Comment by Beep6581** [comment]
@pi99y http://rawpedia.rawtherapee.com/File_Paths#Config
Most likely in `~/.config/RawTherapee5-dev/options`

**Comment by pi99y** [comment]
Thanks @Beep6581

Two **more files exhibiting same problem** (see JPG, original TIFF+pp3 included) (37MB)
https://filebin.net/o2zf1ra4qax54991/highlights_problem.zip

**Options file:**
https://filebin.net/a9rzqvajvj0eznwu/options

Thank you guys for looking into this!

**Comment by pi99y** [comment]
P.S. I tried, same happens if the input TIFF is 16bit. (Above examples are on 8bit TIFF).

**Comment by Beep6581** [comment]
@pi99y please try the following in this order using `201712XX-201801XX agfa-vista400 10_cleaned.tif`:
1. Open that TIFF in the Editor tab, ctrl+s, save to JPG and tick "Save processing parameters with image", then upload that PP3.
2. Turn off CIECAM02 and save to JPG (no need to upload the PP3 nor the image). Does the problem persist?
3. Disable the RGB curves, save, does the problem persist?
4. Even if any of the above worked, delete your options file, then open that TIFF (keep CIECAM02 and RGB curves enabled) and save again, does the problem persist?
5. Compile the latest dev branch following [this](http://rawpedia.rawtherapee.com/Linux) (it's as simple as copying and pasting the correct "Dependencies" section, then copy-pasting the "Automatic way" section) , open the TIFF and save to JPG, does the problem persist? Upload `~/programs/rawtherapee/AboutThisBuild.txt`

**Comment by pi99y** [comment]
1.) 
https://filebin.net/kpd4aqpibir76y3a/201712XX-201801XX_agfa-vista400_10_cleaned.jpg
https://filebin.net/kpd4aqpibir76y3a/201712XX-201801XX_agfa-vista400_10_cleaned.jpg.out.pp3

2.) Aha!
Turning off CIE color appearance Model 2002 **removes the problem!**

3.) CIECAM2 & RGB curves disabled = **also no problem**

4.) Yes, the problem is back.

5.) I will do this step at later time today or tomorrow, can't right now.

Thanks!

**Comment by Beep6581** [comment]
Just co be clear, at step 4 you **closed RT, deleted the options file**, ran RT, opened the TIFF with the same PP3 (**CIECAM02 enabled**), saved to JPG and there were yellow patches? I ask because I noticed that your CIECAM02 settings were different in the options file, still despite using your settings I could not reproduce the problem.

**Comment by pi99y** [comment]
@Beep6581 
On previous attempt I, with the RT running:

- first deleted the options file, 
- disabled both options (CIECAM & RGB Curves) 
- then closed & relaunched the app and 
- re-enabled both options (CIECAM & RGB Curves) 
- saved to JPG.
**= had the problem**

---

Now, I with the RT closed I redid the test like this:
- deleted the options file
- ran RT
- disabled both options (CIE & RGBc)
- closed RT
- deleted the options file
- re-ran RT
- enabled both options
- saved as jpg
**= still same highlights problem**

I will do the step "5" later as promised, I'm currently still busy.

P.S. As noted in step "2." CIECAM seems to be the problematic part.

**Comment by pi99y** [comment]
@Beep6581 

5.) **= no problem**
I did the automatic build, everything went fine.
Started the RT (the recently built one) and disabled & re-enabled both options (CIECAM is now in  different place). Saved to JPG. All is good.

So to me it looks like this is the problem with how the version in the official repository was built (that one also does not include the version details).

AboutThisBuild.txt (formatting messed up, sorry)
```
Version: 5.3-514-gac1238e
Branch: dev
Commit: ac1238e
Commit date: 2018-01-22
Compiler: cc 5.4.0
Processor: x86_64
System: Linux
Bit depth: 64 bits
Gtkmm: V3.18.0
Lensfun: V0.2.8.0
Build type: release
Build flags:  -std=c++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations -O3 -DNDEBUG -ftree-vectorize
Link flags:  -march=native
OpenMP support: ON
MMAP support: ON
```

**Comment by pi99y** [comment]
SEE NEXT COMMENT
---
After I compiled my own version, which meant updating dependencies (I had them installed from the past but they were outdated), even the **repository** version **does not produce the problem any more** even with CIECAM on.

So are those "build" dependencies somehow still used when one installs from the repository? It looks so.

**Comment by pi99y** [comment]
**UPDATE**

I now re-configured (since I guess deleting options file reset that) the "repository version" of RT to **output the JPG in AdobeRGB and the problem is back!**
And... the same problem happens with the self-compiled RT (when the output set accordingly). 

**So the problem is in combination (of CIECAM) with AdobeRGB output.**

**Comment by Benitoite** [comment]
By repository version do you mean the Ubuntu unstable repository, or your build from the official Github repository?

**Comment by pi99y** [comment]
@Benitoite  Yeah sorry, I meant the Ubuntu repository ppa:dhor/myway

And just in case ... I use this AdobeRGB profile file: https://filebin.net/jj3u2sfrzokrdq09/AdobeRGB1998.icc

**Comment by Beep6581** [comment]
@pi99y very good! Now we know that the problem involves CIECAM02 code and is probably related to either the version of a dependency or to the build flags used when compiling that dependency or RT.

> After I compiled my own version, which meant updating dependencies (I had them installed from the past but they were outdated), even the repository version does not produce the problem any more even with CIECAM on.

@Desmis @heckflosse @agriggio @Floessie maybe you can deduce what the problem was? Random guess, maybe related to FFTW3 or LCMS2?

@pi99y I fixed the formatting in your post - you can click the edit button to see how I did it.

**Comment by Beep6581** [comment]
"Great" news, I can reproduce it using your ICC file.

4 days ago I did extract `AdobeRGB.icc` from your sample JPG files but could not reproduce the problem back then. Odd...

... maybe it's only reproducible on Mondays.

**Comment by Beep6581** [comment]
Findings:
1. The problem can be reliably reproduced using default options + neutral + CIECAM02 + the supplied `AdobeRGB1998.icc` or the one from ADOBE.
2. Preferences > Color Management > CIECAM02 > "Use float precision instead of double" has no influence.
3. Your `AdobeRGB1998.icc` does not have the same hash as the one from Adobe http://supportdownloads.adobe.com/thankyou.jsp?ftpID=4075&fileID=3790
However, both lead to identical results with the yellow blob in the lamp and a small green blob above it.
4. RawTherapee's `RT_Medium_gsRGB`, which is [similar](http://rawpedia.rawtherapee.com/Color_Management#Output_Profile) [to](http://rawpedia.rawtherapee.com/Color_Management_addon#Shipped_files) `AdobeRGB1998`, does not exhibit this problem.
5. I can also reproduce it using our shipped ACES and both of our DCI-P3 profiles.

Files: https://filebin.net/yl2akjmzw8j2sq7j

**Comment by pi99y** [comment]
Great, I'm glad we're making progress and seem to be on the right track now.
Thanks again everyone for your work!

I'll keep an eye on the thread and provide testing/files upon your request...

And also thanks for the formatting tip!

**Comment by Beep6581** [comment]
Neutral + CIECAM02 + DCI-P3 output profile:
![amsterdam](https://user-images.githubusercontent.com/5844619/35247260-0a2c9030-ffca-11e7-997e-7f6cfe0dd7d5.jpg)

https://filebin.net/dogpfudtt6jirxm3/amsterdam.jpg.out.pp3

**Comment by Desmis** [comment]
I don't say that there is no problem with Cieacm02. It is well known that there are.

But in this case : Use of "DCI P3" as output profile, with "ciecam enabled" or whith "ciecam disabled", and amsterdam.pef ==> output TIF, there is many many artifacts :)

**Comment by Beep6581** [comment]
Confirmed, the artifacts are there with CIECAM02 disabled using DCI-P3, ACES and AdobeRGB1998, in both JPG and 16-bit TIFF. The RT_* profiles are immune.

**Comment by Beep6581** [comment]
Could not reproduce using 5.3, so this is a regression.

**Comment by agriggio** [comment]
@Beep6581 thanks for the info, I'll take a look

**Comment by agriggio** [comment]
should be fixed now, please test

**Comment by Beep6581** [comment]
Fix confirmed.

**Comment by Beep6581** [comment]
By the way, twice I experienced a complete system slowdown when testing saving this TIFF image to all other formats. Haven't been able to connect with GDB to find out what was going on. The threads were using 100% CPU and I guess I ran out of RAM due to some major leak.

For example I just saved [the TIFF](https://filebin.net/yl2akjmzw8j2sq7j) to JPG, TIFF-8/16 and PNG-8/16, then had to switch to TTY1 to kill RT.
